### PR TITLE
ARROW-1019: [C++] Implement compressed streams

### PR DESCRIPTION
**Pull request by pitrou**
Implement `CompressedInputStream` and `CompressedOutputStream` C++ classes. Tested with gzip, brotli and zstd codecs.

I initially intended to expose the functionality in Python, but `NativeFile` expects a `RandomAccessFile` in read mode (rather than a mere `InputStream`).

### CODE REVIEW

**Comment by wesm** [comment]
Will review this today

File: cpp/src/arrow/io/io-compressed-test.cc [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+  auto compressed_data = MakeRandomData(10000);
+
+  auto buffer_reader = std::make_shared<BufferReader>(Buffer::Wrap(compressed_data));
+  std::shared_ptr<CompressedInputStream> stream;
+  ASSERT_OK(CompressedInputStream::Make(codec.get(), buffer_reader, &stream));
+  std::shared_ptr<Buffer> out_buf;
+  ASSERT_RAISES(IOError, stream->Read(1024, &out_buf));
+}
+
+// NOTE: Snappy doesn't support streaming decompression
```
Comment (wesm) [inline cpp/src/arrow/io/io-compressed-test.cc]: We could (should?) define a framed format of our own devising

**Review by wesm — COMMENTED** [review]
Some minor comments, but this looks great!

File: cpp/src/arrow/io/io-compressed-test.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/io/io-compressed-test.cc]: But is it useful? The main point here is to interact with existing files.

File: cpp/src/arrow/io/compressed.h [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+namespace io {
+
+class ARROW_EXPORT CompressedOutputStream : public OutputStream {
+ public:
+  ~CompressedOutputStream() override;
+
+  /// \brief Create a compressed output stream wrapping the given output stream.
+  static Status Make(util::Codec* codec, std::shared_ptr<OutputStream> raw,
+                     std::shared_ptr<CompressedOutputStream>* out);
+  static Status Make(MemoryPool* pool, util::Codec* codec,
```
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/io/compressed.h]: It's more of an argument passing consistency. We generally use `const T&` for immutable arguments, `T*` for mutable arguments, and `const T*` in the rarer case when the input may be null

**Comment by wesm** [comment]
+1. Let me have a look at the appveyor build

**Comment by wesm** [comment]
I'll wait a little while to make sure appveyor looks good then merge

**Pull request closed without merging**